mailpoet/mailpoet
# Add order status change trigger [MAILPOET-4935]

## Description
Adds the order status change trigger.

## Code review notes
This PR introduces subject transformation. Therefore integrations can now register `SubjectTransformer` which `accept()` a certain type of subject and `return()` a certain type using `transform()`.

### The new `SubjectTransformerHandler` can 
1. Generate all possible subjects for an automation. It takes the subjects which are provided by the trigger and searches for possible transformations. This is used to be able to add steps to an automation, which do require different subjects than the trigger provides but are available thanks for transformations. In the ValidStepOrderRule this functionality replaces now the `collectSubjectKeys()` method.
2. Generate a transformation chain, which then transforms a subject A into a subject B. This transformation takes place, when we search for the required subjects in an automation in order to execute a specific step.

Why a chain? Imagine a trigger provides a `subject_a` and we need a `subject_c`, but there is no straight transformation. What we have is an `AtoB` and a `BtoC` transformer. The chain enables us to transform now `subject_a` to `subject_c` by using both transformers in an orderly fashion.

### The new WooCommerce integration
Now we can have a order status change trigger, which does not have a subscriber subject. With the new transformer from order to subscriber, we can transform these subjects when needed.

The review remarks of the [old PR](https://github.com/mailpoet/mailpoet/pull/4748) have been taken care of.

## QA notes

Besides checking the trigger if it works as expected, please check also if the "Run this automation only once per subscriber" works as expected.

## Linked PRs
https://github.com/mailpoet/mailpoet-premium/pull/686

## Linked tickets

 [MAILPOET-4935]

## After-merge notes

Do not merge before transactional emails have been merged. Ping @websupporter when in doubt.

[MAILPOET-4935]: https://mailpoet.atlassian.net/browse/MAILPOET-4935?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
